Readme.md
# RoRToolkit

Rigs of Rods Toolkit, abbreviated RoRToolkit, is a multi-purpose content creation/modification tool for Rigs of Rods.

Wikipage: http://www.rigsofrods.com/wiki/pages/RoRToolkit

## Features:

Editing maps (.terrn format)
Editing map objects (.odef format)
Editing vehicles (.truck format)

## How to start

Startup settings dialog
NOTE: Currently, only Windows version is supported. Linux status is unknown. Get in touch with developers for more info.

Download from forum thread [RoRToolkit is coming back to life](http://www.rigsofrods.com/threads/120839)

Installation: Just unpack the archive. No installation is required.


Main screen
Startup: Doubleclick "rortoolkit.bat". You'll be shown a "Startup settings dialog". Perform the following steps:

* Step1: Select RoR installation directory. At the moment, it MUST be an installation of RoR 0.38 - other versions will either not be detected at all or won't start properly
* Step2: Select RoR home directory. On windows, this is [MyDocuments/Rigs of Rods *]. RoRToolkit will detect all available directories and show a selection list. Choose any. However, RoRToolkit only supports editing maps in .terrn format, which is only supported in 0.38 and older versions.
* Step3: Choose a rendering system. On windows, both work fine - choose any.
* Step4: The "start" button will become enabled once you complete steps 1+2. Click it.
Then, RoRToolkit starts up the 3D engine, loads resources (this may take a while, it loads everything, users already complained) and you'll land in the main screen.

## How to edit maps

How to open map for editing
Use "Tree view" panel:

Doubleclick the "Terrains" directory to expand it
Right-click the desired terrain -> a popup menu will appear. Select "Edit in terrain editor"
This is an older (the "installation" part is out of date) [youtube video tutorial on map editing / adding road](https://www.youtube.com/watch?v=Ewhd65YcFlw&feature=youtu.be&t=1m8s)

## Get Involved

See there: [RoRToolkit Developer](http://www.rigsofrods.com/wiki/pages/RoRToolkit_Developer). Also suggest ideas.

ChangeLog.txt

--------------------------------------------------------------------------------
                            RoR Toolkit ChangeLog
--------------------------------------------------------------------------------

NEXT:
	* Removed bitmap skins - windows are plain "tool panels" now.
	* Mouse translation arrows now work properly (like in Blender).
	* Startup settings dialog now lets user choose RoR home directory
	  (previously hardcoded to "My Documents/Rigs of Rods 0.38").
	* Error handling: added popup dialogs displaying info in case of
	  crash or failure to start.

--------------------------------------------------------------------------------
    2016-01-06: RoR Toolkit ressurected after almost 5 years of inactivity.                      
--------------------------------------------------------------------------------

Changes Made By Max98 31/05/2011

- Compitable with 0.38.X

Changes made by Lepes

- preload system with rotation


before release nov 2009:
- hide paste button / find object
- fixed truckshop 7.2 section
- fixed bookmark camera that aren't show until you move them
- fixed rotation axis colour and behaviour
- fixed file loggin.ini encoding as ANSI file since it could contain non-ascii
  characters in windows username or paths
- fix above on all toolkit files !!!! argggg!!!
- fix starter dimension window


Thanks to all invited members, they help me a lot suppling ideas and meshes 
  (Sneer, dieselfuelonly, Lifter...)

2008-09-07
- Terrain Editor:
	+ Start Character Position, camera Start Position: just move red meshes 
	    to the desired position
	
	+ Object Inspector that allow user to type coords and rotation angles. 
	  You can also add a comment to every object (useful to describe the first
	    and last road objects, for example: "begin road to castle city")
	  Comments may be multilinea but if you use "/", it will be deleted.
	  ENTER key must be the last key to press, 
	    because it is when text and comments are saved.
	
	+ Contigous Block Mode: The first block is added like you did it before, 
	    that is: click on terrain ground, then use right button click 
		on Object Tree
	  to add the object to the terrain and now, click the object just added, 
	    while it is highlighted press SPACE BAR key, you would see another piece 
	  automatically added updating X, Y, Z coords and copying RotX, RotY, RotZ.
	  
	  Keep in mind that SPACE BAR will add the selected item 
	    in Object Tree next to the selected one on 3D-Window.
	  3D-Window must be focused (just right click in the window) before 
	    pressing SPACE.
	
	+ alphabetically Ordered Object Tree for each category (names are changed 
	  to lowercase to maintain a consistent ordered items)
	
	+ Support Multiple SpawnZones on Terrain.
	+ keys UIO rotate to create ascend/descend ramp. keys JKL  rotate to create
	  corners keys M<comma><dot> rotate on X axis (banked curves??)
	+ Skins support for Object Tree, Object Inspector, 
	  Object Preview and Map Options
	+ Control Key while selecting a piece and dragging = freedom movement 
	  (the piece should be unselected, then keep CTRL pressed while selecting)
	+ Q key attach camera to selected object so, you can rotate camera 
	  with AWSDFV keys around the selected object.
	+ avoid to select deleted/invisible objects.
	+ Improved singleton pattern
	+ If no Camera/Character/Truck Start Position are supplied on
	  the .terrn file or its coordenates are (0,0,0) 
		they are setup on the middle of the map 
		  at terrain Height or water Level (max value)
		 
- Odef Editor
	- fixed bug at initialization
	- change camera position
	- Q key allow rotate the object at centered view (just as terrain editor)


- Truck Editor	
	- fixed bug at initialization  
	- fixed multiples optional values in beams sections and nodes section, 
	  for example "ri" beams
	- Truck file parser updated to wiki 2009-01-21
	- Added new bugs ;-)

- Object Preview, Map Preview
    - Support MouseWheel

- Main configuration:
	- starter.py - checkRoRDir:(minor change) some new strings to show the user
	  that RoR binary executable had not been found and the cause.  	  

	   

modtool.py
#Thomas Fischer 31/05/2007, thomas@thomasfischer.biz
import sys, os, os.path, shutil

sys.path.append(os.path.join(os.path.dirname(os.path.abspath(__file__)), "lib"))
from ror.logger import log
from ror.settingsManager import getSettingsManager

TEMPDIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), "temp")

def removetemp(reporterrors=True):
    if os.path.isdir(TEMPDIR):
        try:
            shutil.rmtree(TEMPDIR)
            #os.rmdir(TEMPDIR)
        except Exception, err:
            if not reporterrors:
                return
            log().error(str(err))
            log().error("could not remove temporary diretory: %s! please delete by hand." % TEMPDIR)
            sys.exit(1)

def ExtractToTemp(filename):
    file, extension = os.path.splitext(filename)
    removetemp(False)
    os.mkdir(TEMPDIR)
    if extension.lower() == ".rar":
        import UnRAR        
        dst = os.path.join(TEMPDIR, os.path.basename(filename))
        shutil.copyfile(filename, dst)
        os.chdir(TEMPDIR)
        UnRAR.Archive(os.path.basename(filename)).extract()
        
        # remove .rar file instantly
        os.unlink(os.path.join(TEMPDIR, os.path.basename(filename)))
        
        # change back to current path
        os.chdir(os.path.dirname(os.path.abspath(__file__)))
        return True
    elif extension.lower() == ".zip":
        import UnZIP
        UnZIP.unzip(filename, TEMPDIR)
        return True
    else:
        log().info("copying "+filename+" to "+os.path.join(TEMPDIR, os.path.basename(filename)))
        shutil.copyfile(filename, os.path.join(TEMPDIR, os.path.basename(filename)))
    return False

def installfile(maintarget, srcfile, dryrun):
    file, extension = os.path.splitext(maintarget)
    rorpath = getSettingsManager().getSetting("RigsOfRods", "BasePath")
    
    if extension in ['.truck', '.load']:
        path = os.path.join(rorpath, "data", "trucks")
    elif extension in ['.terrn']:
        path = os.path.join(rorpath, "data", "terrains")
    else:
        path = rorpath
    if dryrun:
        log().info("would install %s to %s" % (os.path.basename(srcfile), path))
    else:
        log().info("installing %s to %s" % (os.path.basename(srcfile), path))
        shutil.copyfile(srcfile, os.path.join(path, os.path.basename(srcfile)))

def searchFile(filename, top):
    for root, dirs, files in os.walk(top, topdown=False):
        if filename in files:
            return os.path.join(root, filename)
    return None

def getTargets(verbose):
    import ror.depchecker
    dc = ror.depchecker.RoRDepChecker(TEMPDIR, "getfiles", "", False)
    targets = []
    for file in dc.files:
        filename, extension = os.path.splitext(file)
        if extension.lower() in ['.truck', '.terrn', '.load']:
            targets.append(os.path.basename(file))

    validtargets = []
    invalidtargets = []
    if len(targets) == 0:
        log().info("### no targets found")
        return validtargets, invalidtargets
    
    log().info("### found %d targets, checking them separatly now" % len(targets))
    for target in targets:
        log().info("### checking target %s..." % target)
        dc = ror.depchecker.RoRDepChecker(TEMPDIR, "dtree", target, verbose)
        if dc.everythingfound:
            validtargets.append(target)
        else:
            invalidtargets.append(target)
    return validtargets, invalidtargets

def getRoRMods(verbose):
    import ror.depchecker
    rorpath = getSettingsManager().getSetting("RigsOfRods", "BasePath")
    dc = ror.depchecker.RoRDepChecker(rorpath, "getfiles", "", False)
    targets = []
    for file in dc.files:
        filename, extension = os.path.splitext(file)
        if extension.lower() in ['.truck', '.terrn', '.load']:
            targets.append(os.path.basename(file))
    newtargets = []
    md5s = dc.readMD5File()
    for target in targets:
        if not os.path.basename(target) in md5s.keys():
            newtargets.append(target)
    return newtargets


def work(mode, targetfile, verbose, dryrun, installtarget=None):
    log().info("### modinstaller started with %s, %s" % (mode, targetfile))
    if mode == "install":
        filename = os.path.abspath(targetfile)
        ExtractToTemp(targetfile)
        target = installtarget
        log().info("### validating target ...")
        import ror.depchecker
        dc = ror.depchecker.RoRDepChecker(TEMPDIR, "dtree", target, verbose)
        if dc.invalid:
            log().error("### target invalid!")
            log().info("### please use the list mode to get valid targets")
            usage()
        
        log().info("### target valid!")
        #print dc.dstree
        installcounter = 0
        for file in dc.dstree:
            filename = file['filename']
            filenamefound = searchFile(filename, TEMPDIR)
            if filenamefound is None:
                log().error("File %s not found in %s!" % (filename, TEMPDIR))
                sys.exit(1)
            installfile(target, filenamefound, dryrun)
            installcounter += 1
        if dryrun:
            log().info("### would install %d files." % installcounter)
        else:
            log().info("### %d files installed, finished!" % installcounter)
        removetemp(False)
        return [target]    

    elif mode == "installall":
        filename = os.path.abspath(targetfile)
        ExtractToTemp(targetfile)
        validtargets, invalidtargets = getTargets(verbose)
        log().info("### installing %d found modifications:" % (len(validtargets)))
        installcounter = 0
        for target in validtargets:
            log().info("### installing modification '%s'" % target)
            import ror.depchecker
            dc = ror.depchecker.RoRDepChecker(TEMPDIR, "dtree", target, verbose)
            if dc.dstree is None:
                log().error("no dependenytree for File %s!" % (filename))
                continue
            for file in dc.dstree:
                filename = file['filename']
                filenamefound = searchFile(filename, TEMPDIR)
                if filenamefound is None:
                    log().error("File %s not found in %s!" % (filename, TEMPDIR))
                    sys.exit(1)
                installfile(target, filenamefound, dryrun)
                installcounter += 1
        if dryrun:
            log().info("### would install %d files." % installcounter)
        else:
            log().info("### %d files installed, finished!" % installcounter)
        removetemp(False)
        return validtargets    
    
    elif mode in ["list", "listall"]:
        filename = os.path.abspath(targetfile)
        ExtractToTemp(targetfile)
        validtargets, invalidtargets = getTargets(verbose)
        if mode == "listall":
            if len(invalidtargets) > 0:
                log().info("broken modifications found:")
                for v in invalidtargets:
                    log().info("  %-20s" % v)
                log().info("use the --verbose flag to find the missing files!")
            else:
                log().info("no broken modifications found")

        if len(validtargets) > 0:
            log().info("installable modifications found:")
            for v in validtargets:
                log().info("  %-20s" % v)
        else:
            log().info("no installable modifications found! :(")           
        # todo : remove workaround!
        removetemp(False)
        return validtargets

    elif mode in ["listinstalled"]:
        targets = getRoRMods(verbose)
        if len(targets) > 0:
            log().info("### Found Mods:")
            for target in targets:
                log().info("  "+target)
        else:
            log().info("### No Mods found!")
            
    if mode in ["uninstall"]:
        rorpath = getSettingsManager().getSetting("RigsOfRods", "BasePath")
        log().info("### validating target ...")
        import ror.depchecker
        dc = ror.depchecker.RoRDepChecker(rorpath, "dtree", targetfile, verbose)
        if dc.invalid:
            log().error("### target invalid! (Target not found)")
            log().info("### please use the 'listinstalled' mode to get valid uninstallation targets")
            return None
        
        log().info("### target valid!")

        #print dc.dstree
        newtargets = []
        md5s = dc.readMD5File()
        for file in dc.dstree:
            filename = file['filename']
            if not os.path.basename(filename) in md5s.keys():
                newtargets.append(filename)
        log().info("### removed %d files from dependency tree." % (len(dc.dstree)-len(newtargets)))
        #print newtargets
        if dryrun:
            log().info("### would uninstall %d file(s):" % len(newtargets))
        else:
            log().info("### uninstalling %d file(s):" % len(newtargets))
        for target in newtargets:
            filenamefound = searchFile(target, rorpath)
            if filenamefound is None:
                log().error("### File not found: %s" % target)
                continue
            log().info("   %s" % filenamefound)
            if not dryrun:
                os.unlink(filenamefound)

    return None

def usage():
    print "usage (general): %s <mode> <additionaloptions> [--verbose] [--dryrun]" % (os.path.basename(sys.argv[0]))
    print "list <filename>"
    print " list all found and valid modifications in filename"
    print ""
    print "listall <filename> --verbose"
    print " list all found modifications in filename (valid and invalid)"
    print ""
    print "installall <filename> --verbose --dryrun"
    print " install all found modifications in filename"
    print ""
    print "install <filename> <modification> --verbose --dryrun"
    print " install a certain modifications in filename (valid and invalid)"
    print ""
    print "listinstalled"
    print " lists all installed RoR Mods"
    print ""
    print "uninstall <modname> --verbose --dryrun"
    print " uninstalls a mod"
    print ""
    print "notes: the --verbose option is optional to increase the output for debugging"
    print "       the --dryrun option is optional and prints out what would be done"
    sys.exit(0)

def main():

    # check for valid RoR Directory!
    import ror.settingsManager
    rorpath = ror.settingsManager.getSettingsManager().getSetting("RigsOfRods", "BasePath")
    if not os.path.isfile(os.path.join(rorpath,"RoR.exe")):
        import ror.starter
        ror.starter.startApp()

    if len(sys.argv) < 2:
        usage()

    mode = sys.argv[1]
    if not mode in ['list', 'listall', 'install', 'installall', 'listinstalled','uninstall']:
        usage()
    if len(sys.argv) < 4 and mode in ['install']:
        usage()
            
    if mode in ['list', 'listall', 'install', 'installall']:
        targetfile = sys.argv[2]
        if not os.path.isfile(targetfile):
            log().error("%s is not a valid target filename!" % targetfile)
            usage()
    elif mode in ['listinstalled']:
        targetfile = rorpath
    elif mode == 'uninstall':
        targetfile = sys.argv[2]
    


    # get optional flags
    verbose = False
    dryrun = False
    for option in sys.argv:
        if option == "--verbose":
            verbose = True
        if option == "--dryrun":
            dryrun = True

    # Import Psyco if available
    try:
        import psyco
        psyco.full()
    except ImportError:
        pass
        
    if len(sys.argv) == 4:
        installtarget = sys.argv[3]
    else:
        installtarget = None
    work(mode, targetfile, verbose, dryrun, installtarget)

    

if __name__=="__main__":
    main()